csharp/ql/src/semmle/code/csharp/TypeRef.qll

```diff
@@ -0,0 +1,73 @@
+/**
+ * INTERNAL: Do not use.
+ * Provides support for type-references.
+ */
+
+import csharp
+
+/**
+ * INTERNAL: Do not use.
+ * Gets a type reference for a given type `type`.
+ * This is used for extensionals that can be supplied
+ * as either type references or types.
+ */
+@type_or_ref getTypeRef(@type type) {
+  result = type
+  or
+  type = result.(TypeRefs::TypeRef).getCanonicalType()
+}
+
+private module TypeRefs {
+  /**
+   * A typeref is a reference to a type in some assembly.
+   * Often, a type can be present in multiple assemblies.
+   */
+  class TypeRef extends @typeref {
+    string getName() { typerefs(this, result) }
+
+    string toString() { result = getName() }
+
+    Type getAType() { typeref_type(this, result) }
+
+    /**
+     * Gets the "canonical type" represented by a typeref.
+     * This is because a typeref can reference multiple types, or a type with the same
+     * fully qualified name can exist in multiple assemblies.
+     * The canonical type is an arbitrarily chosen type from the list of candidate types.
+     */
+    Type getCanonicalType() {
+      result = this.getAType() and
+      isCanonicalType(result)
```

Comment:
Since each type can be mapped by at most one `TypeRef`, I think this i slightly clearer (and `isCanonicalType` can be removed):
```
not locationIsBetter(canonicalTypeLocation(result), canonicalTypeLocation(this.getAType()))
```